src/consensus/tx_verify.cpp

```diff
@@ -181,25 +185,167 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
     }
 
     // Check for duplicate inputs - note that this check is slow so we skip it in CheckBlock
-    if (fCheckDuplicateInputs) {
-        std::set<COutPoint> vInOutPoints;
-        for (const auto& txin : tx.vin)
-        {
-            if (!vInOutPoints.insert(txin.prevout).second)
-                return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-duplicate");
-        }
-    }
-
-    if (tx.IsCoinBase())
-    {
+    if (tx.IsCoinBase() ) {
         if (tx.vin[0].scriptSig.size() < 2 || tx.vin[0].scriptSig.size() > 100)
             return state.DoS(100, false, REJECT_INVALID, "bad-cb-length");
-    }
-    else
-    {
-        for (const auto& txin : tx.vin)
-            if (txin.prevout.IsNull())
+    } else if (!fCheckDuplicateInputs || tx.vin.size() == 1){
+        for (const auto& txin : tx.vin) {
+            if (txin.prevout.IsNull()) {
                 return state.DoS(10, false, REJECT_INVALID, "bad-txns-prevout-null");
+            }
+        }
+    } else{
+        // This duplication checking algorithm uses a probabilistic filter
+        // to check for collisions efficiently.
+        //
+        // This is faster than the naive construction, using a set, which
+        // requires more allocation and comparison of uint256s.
```

Comment:
See https://github.com/bitcoin/bitcoin/pull/14397 for a more obvious version that is just the obvious thing.

For std::unordered_set, I'm clocking much worse performance for DeserializeAndCheckBlockTest and 2x worse performance for DuplicateInputs.

Reply:
@leishman I tried a few alternatives:
* Master: 13.6 ms
* #14397: 6.3 ms
* Using a sorted vector with SipHash'ed prevouts: 3.7 ms
* This PR: 2.7 ms

Reply:
I like sorting the siphash'd prevouts -- I'm guessing you then do the expensive check if that collides? Or are you tracking the pointers when you sort too?

Reply:
@JeremyRubin Yeah, just delay the expensive check until after the cheap check fails. I haven't PR'ed that because I don't have a production version, and I don't want to overload reviewers with a series of PRs without even knowing if we want to increase complexity here at all.

Reply:
For the expensive check, you can still do it in O(n) per colliding entry FYI, which is less expensive than doing the full O(n log n) expensive check given that we don't expect colliding entries without a duplicate.